### ISSUE TITLE
bug: enabling fzf extra via config doesn't work (install telescope specs)

### ISSUE THREAD

**Issue by mirsella** [LazyVim/LazyVim#3626]
### Did you check docs and existing issues?

- [X] I have read all the LazyVim docs
- [X] I have searched the existing issues of LazyVim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.11.0-dev-226+g721551210

### Operating system/version

Linux main 6.9.3-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 31 May 2024 15:14:45 +0000 x86_64 GNU/Linux

### Describe the bug

when launching lazyvim with the fzf extra enabled from the config like so
`    { import = "lazyvim.plugins.extras.editor.fzf" },`
i get the following warning message
```
   Warn  08:02:04 PM notify.warn LazyVim `LazyVim.pick`: picker already set to `telescope`, ignoring new picker `fzf`
```
fzf is not used, fzf-lua not even installed, and the telescope specs are installed instead.
Checking after with :lazyextras the fzf extra is marked as installed, while it's not

enabling the extra with :lazyextras and everything works fine

### Steps To Reproduce

use the fzf extra via config and launch lazyvim

### Expected Behavior

fzf is used like previously

### Repro

repro doesn't work for me, as nothing it installed except lazyvim, lazy.nvim and tokyonight. not all the plugins from lazyvim.
no :lazyextras, only :lazy. can't manage to install any extra in the config. no keymaps.

thanks !

**Comment by dpetka2001** [comment]
Are you sure you don't have some kind of Telescope spec in your personal config?

**Comment by mirsella** [comment]
@dpetka2001 my bad i guess i missed my commands and the repro.lua actually doesn't work.
i can't even manage to specify a extra ? The repro.lua from the first post doesn't even install fzf, or any other extra i put in there.
how can i add a extra to the repro.lua ?

for you question, i removed all telescope config to test, and the message was still there, so i'm sure it's not my config.

**Comment by dpetka2001** [comment]
Strange I've been using it since it got merged and I don't have such a problem. I remember reading that when you import it via `import = "lazyvim.plugins.extras.editor.fzf"` in your `lazy.lua` then the Telescope spec will still be parsed, but only one of the 2 will be enabled. I enabled fzf-lua via `LazyExtras`. Can you try that (if you haven't done so already)?

**Comment by mirsella** [comment]
using the repro.lua, like i said it's bugged and the command :LazyExtras doesn't even exist.

using my personal config, removing the `import [...] editor.fzf` and adding it via lazyextras works.

the problem is indeed the `{ import = "lazyvim.plugins.extras.editor.fzf" },` which install the telescope specs.

**Comment by mirsella** [comment]
i've updated the main post

**Comment by dpetka2001** [comment]
The following `repro.lua` will work correctly
```lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  { "LazyVim/LazyVim", import = "lazyvim.plugins" },
  { import = "lazyvim.plugins.extras.editor.fzf" },
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```
That doesn't solve the problem with fzf-lua. Something is going on with `LazyVim.has_extra`. Unfortunately, I'm not that well-versed in `lazy.nvim`'s internals. Maintainer will know better.

**Comment by mirsella** [comment]
thanks this repro.lua works, because it use `{ "LazyVim/LazyVim", import = "lazyvim.plugins" }` while the default just has `  "folke/LazyVim",`. i'll open a pr for the issue template

**Comment by dpetka2001** [comment]
So, `require("lazy.core.config").spec.modules` only prints out `lazyvim.plugins` at the parsing phase. That leads to the import of fzf-lua to be always `false` and the telescope one to be `true`. When then from `lazy.lua` the `import = "lazyvim.plugins.extras.editor.fzf"` is evaluated the telescope spec has already been enabled and registered a picker and that's why we see the message that you mention. I still fail to think of a good way to get around this due to my lack of knowledge. I wonder if there's a way for `LazyVim.has_extra` to get the imported modules after the parsing phase of all modules has been completed and check based on that.

**Comment by rgzntrade** [comment]
> 我有同样的问题

+1

**Comment by rgzntrade** [comment]
As a temporary solution, set the local telescope configuration of lazyvim to enabled= false,. Then recompile lazyvim, the problem is solved temporarily.

**Comment by konosubakonoakua** [comment]
same here.
My extras config for fzf
```lua
  { import = "lazyvim.plugins.extras.editor.fzf" },
  { "junegunn/fzf", build = "./install --bin" },
  {
    "ibhagwan/fzf-lua",
    optional = true,
}
```
not working

**Comment by konosubakonoakua** [comment]
> So, `require("lazy.core.config").spec.modules` only prints out `lazyvim.plugins` at the parsing phase of the `lazyvim.plugins.editor`. That leads to the import of fzf-lua to be always `false` and the telescope one to be `true`. When then from `lazy.lua` the `import = "lazyvim.plugins.extras.editor.fzf"` is evaluated the telescope spec has already been enabled and registered a picker and that's why we see the message that you mention. I still fail to think of a good way to get around this due to my lack of knowledge. I wonder if there's a way for `LazyVim.has_extra` to get the imported modules after the parsing phase of all modules has been completed and check based on that.

How about this way
```lua
  { import = "lazyvim.plugins.extras.editor.fzf" },
  { import = "lazyvim.plugins.extras.editor.telescope", enabled = false},
```

**Comment by folke** [comment]
Just use `:LazyExtras`?
I've changed it so that we no longer have to process the full telescope spec when you have fzf-lua enabled.

That only works properly if you added fzf-lua with `:LazyExtras`.

**Comment by mirsella** [comment]
well, if :LazyExtras is the recommanded / supported way, it should be said in the docs.
the config option should be deprecated if it doesn't work, removed from the docs etc.

personally I like the config option because I can put in my config (which is synced between computer) which extra is enabled, and I have a lot of them.
each time I setup my dotfiles I don't want to have to run lazyextras and remember which one I had, while it could just work by setting it in the config

**Comment by rgzntrade** [comment]
> ```lua
> { import = "lazyvim.plugins.extras.editor.telescope", enabled = false},
> ```
I also commented out these lines from telescope.lua
-- if not LazyVim.pick.register(picker) then
--   return {}
-- end

I installed it the same way.
{ import = "lazyvim.plugins.extras.editor.fzf" },

**Comment by mirsella** [comment]
@EraPartner didn't know about lazyvim.json, I'll check it out ! it'll will probably work for me, thanks !

**Comment by mirsella** [comment]
i guess we can close this, but we should update the docs to deprecate the old method, as it's causing problems and is not supported anymore

**Comment by dpetka2001** [comment]
@mirsella 

> well, if :LazyExtras is the recommanded / supported way, it should be said in the docs.
the config option should be deprecated if it doesn't work, removed from the docs etc.

The import config option is something that comes from `lazy.nvim` itself. It just doesn't work correctly in this specific case, because Telescope was also moved to the Extras and the `import` is happening in `lazyvim.plugins.editor`. It still works for other cases. So, the old method still works, but requires a bit of knowledge about what you're doing (i.e with the imports in your config you should also be vigilant of the order you put the imports, because some Extras need to be loaded first). The docs for the Extras first mention as an option to install the Extra via LazyExtras and then mention the config as alternative. For the specific case of fzf/telescope, maybe it could indeed be mentioned that LazyExtras should be the preferred way. 

But it's not like the config option is deprecated or not supported anymore in general. It just doesn't work correctly for this specific change, so any change in the docs should be made in the page of fzf/telescope Extras.

**Comment by mirsella** [comment]
Thanks for the explanations.
the only place where the config option is still remaining is the starter template and on https://www.lazyvim.org/configuration/lazy.nvim which just show the starter template.

i'll open a pr on the starter template comments to better reflect the status of the config import, as for a long term user which was used to config option, and for a new user that see the starter template, it's confusing, and will affect everyone that use it

so in conclusion,
maybe a special warning on the fzf extra page to warn that enabling it via config option doesn't work ?
or a general tips that the setting extras from the import config can cause problem like for fzf ?

**Comment by dpetka2001** [comment]
In my personal opinion I wouldn't mess with LazyVim starter template, as that is still valid. I believe in this specific case, a special warning on LazyVim website in the page of fzf Extra about explicitly enabling it via LazyExtras would be preferred.

**Comment by mirsella** [comment]
Still valid until someone tries to add fzf haha.
The comment about extras has been removed in the starter template, to avoid problem and because :lazyextra is preferred instead.

i'll try to do a pr for the docs

**Comment by Drewtopia** [comment]
When I enabled `editor.fzf` via :LazyExtras, I am seeing the following errors

![CleanShot 2024-06-14 at 22 38 05@2x](https://github.com/LazyVim/LazyVim/assets/1256558/9a717396-34a4-45c2-bff3-98c34e84ea88)
![CleanShot 2024-06-14 at 22 54 23@2x](https://github.com/LazyVim/LazyVim/assets/1256558/65677abb-90fd-415f-84ab-dc9edca0e125)

**Comment by mirsella** [comment]
@Drewtopia that's a different issue, you can open a new one for it to be treated

**Comment by folke** [comment]
@Drewtopia don't open an issue here. Check the fzf-lua github page first for related issues and open one there if you couldn't fix it yourself

**Comment by ibhagwan** [comment]
@Drewtopia, that’s the same issue as https://github.com/ibhagwan/fzf-lua/issues/1243, usually indicates an issue with your neovim setup (specially access to runtime folder) which makes `vin.fn.serverstart` fail, also look into the upstream issue I linked it will have some pointers for fixing this.

**Comment by hardyweb** [comment]
> When I enabled `editor.fzf` via :LazyExtras, I am seeing the following errors
> 
> ![CleanShot 2024-06-14 at 22 38 05@2x](https://github.com/LazyVim/LazyVim/assets/1256558/9a717396-34a4-45c2-bff3-98c34e84ea88) ![CleanShot 2024-06-14 at 22 54 23@2x](https://github.com/LazyVim/LazyVim/assets/1256558/65677abb-90fd-415f-84ab-dc9edca0e125)

https://dev.to/hardyweb/lazyvim-version-14x-in-wsl-2n8p

**Comment by surajsharma** [comment]
> In my personal opinion I wouldn't mess with LazyVim starter template, as that is still valid. I believe in this specific case, a special warning on LazyVim website in the page of fzf Extra about explicitly enabling it via LazyExtras would be preferred.

your very public opinion voiced here is the antithesis of the whole concept of configuration

**Comment by folke** [comment]
@surajsharma you either use `:LazyExtras` or you deal with correct import ordering yourself.

It's highly recommended **not** to fiddle with your lazy config regarding extras.

Ordering of imports matters and if you do it manually, then you're on your own.
